### ISSUE TITLE
using BERT instead of SpaCy

### ISSUE THREAD

**Issue by abecode** [abecode/teaching_reviews#13]
see if BERT gets better performance than spacy.  To do this will require looking at redact separately because that is the label that overlaps with other labels.

The hypothesis is that the span labeling can be useful for annotation but for actual classification it is better to use a traditional BERT non-overlapping chunk model

**Comment by fennialee** [comment]
This notebook use DistillBert for annotation. Note that a huggface account and wandb API is required for training this model.
https://colab.research.google.com/drive/1zQ6kkT8wW5EHt38IjUir-VApNN-shBPe?usp=sharing
The model has good performance on POSITIVE and NEGATIVE, acceptable performance on SUGGESTION, however, it has bad performance on REDACT.

**Comment by mengyuan5483** [comment]
https://colab.research.google.com/drive/1uOjOc_UDrspyHw-sPFhn3ZvhzxhJOGLk?usp=sharing
This notebook is Fine-tuning BERT on Teaching Review Classification. The performance is good on label "suggestion". I cannot figure out why the column names are missing the first letters.

**Comment by mengyuan5483** [comment]
I have fixed that bug about the first letter missing, it looks like there is something wrong about seqeval.https://colab.research.google.com/drive/1uOjOc_UDrspyHw-sPFhn3ZvhzxhJOGLk?usp=sharing

**Comment by abecode** [comment]
Here's a [notebook https://colab.research.google.com/drive/17T0nUW4CUK0W2IUCxRMhFSCDPHcV7srk?usp=sharing](https://colab.research.google.com/drive/17T0nUW4CUK0W2IUCxRMhFSCDPHcV7srk?usp=sharing) that I think fixes some issues in the previous notebooks (but not all).

- first, it looks like the previous notebooks don't have a proper train/dev/test set split.  I rewrote that to match what I was doing with spacy
- second, it doesn't make sense to include REDACT in the chunking/NER approach, so I took out REDACT.  We can do a separate pass with REDACT
- third, I don't think that the span annotations were filtered to the >= 3 agreement level.

Not surprisingly, when we remove REDACT, the performance improves a lot because it is impossible to have overlapping spans with BERT type sequence labeling models.

```
              precision    recall  f1-score   support

    NEGATIVE     0.8144    0.7577    0.7850       359
           O     0.9964    0.9909    0.9936     13778
    POSITIVE     0.6448    0.8955    0.7498       440
  SUGGESTION     0.6634    0.4945    0.5666       271

    accuracy                         0.9733     14848
   macro avg     0.7798    0.7846    0.7737     14848
weighted avg     0.9755    0.9733    0.9736     14848
```

**Comment by abecode** [comment]
Here's [another notebook to compare BERT and DistilBERT](https://colab.research.google.com/drive/1U_B6SD_U0A6qmp6YR6KuzrDBwcwXVJhB?usp=sharing).  Actually using 4 epochs for both gives the same results!  It seems surprising, but if it's correct it at least shows that the models generalize in the same way, repeatably.  If anyone has spare time, they might look at varying the number of epochs of training.

**Comment by abecode** [comment]
Here's my notebook for IOB with BERT and DistilBERT: [https://colab.research.google.com/drive/1umSJjR2zfZfRs2NQFyXQ4_xkK4YWKSze?usp=sharing](https://colab.research.google.com/drive/1umSJjR2zfZfRs2NQFyXQ4_xkK4YWKSze?usp=sharing)

It is a bit messy and it changed the earlier learning rate to 10 epochs.

**Comment by fennialee** [comment]
This notebook is for training with DistilBERT.
https://colab.research.google.com/drive/1xPH9glGXDvQti6WG6x306ehXm5Ckyc5x?usp=sharing
The training parameter ( num of epoch, batch size, learning rate) are set the same as training with BERT, so we can do a side by side comparison.
